R/future_promise.R

```diff
@@ -281,7 +300,194 @@ future_promise_queue <- local({
+#' @return a [promise()] object that will return the result of the calculated `expr`.
+#' @examples
+#' \donttest{local({
+#'   ## Setup ##
+#'
+#'   # Reset future plan on exit
+#'   old_plan <- future::plan()
+#'   on.exit({future::plan(old_plan)}, add = TRUE)
+#'
+#'   # Set up a plan with 2 future workers
+#'   future::plan(future::multisession(workers = 2))
+#'
+#'   # `cat()` to file so that the console does not display confusing results
+#'   log <- tempfile()
+#'   on.exit({unlink(log)}, add = TRUE)
+#'   start <- Sys.time()
+#'   cat_ex <- function(...) {
+#'     msg <- paste0(..., "; ", sprintf("%.2f", difftime(Sys.time(), start, units = "secs")))
+#'     message(msg)
+#'     cat(msg, "\n", sep = "", file = log, append = TRUE)
+#'   }
+#'
+#'   # This function will print every `delay` seconds
+#'   # The thing to notes is that the function will not execute unless the main R session is free
+#'   # We should expect to see `print_every()` statements interleaved with `future_promise()` calls
+#'   # We should NOT expect to see `print_every()` statements
+#'   #   while `future::future()` is blocking the main R session
+#'   print_every <- function(i = 0, max = 5 / delay, delay = 0.25) {
+#'     if (i > max) return()
+#'     cat_ex("print_every(): ", i, "/", max)
+#'     # Print again, later
+#'     later::later(function() { print_every(i + 1, max = max, delay = delay) }, delay = delay)
+#'   }
+#'
+#'   # Act as if the main R session is free for the next `timeout` seconds
+#'   run_ex_now <- function() {
+#'     while (!later::loop_empty()) {
+#'       later::run_now()
+#'       Sys.sleep(0.01)
+#'     }
+#'     message("done with run_ex_now()")
+#'   }
+#'
+#'   # Consistent printing method for the example below
+#'   print_ex <- function(x) {
+#'     promise_all(.list = x) %...>%
+#'       {
+#'         cat("\n")
+#'         print(.);
+#'         cat("\nlog:\n", paste0(readLines(log), collapse = "\n"), "\n", sep = "")
+#'       }
+#'   }
+#'
+#'
+#'   ## Example ##
+#'
+#'   # Execute 10 `future_promise()` calls.
+#'   # In this example, we expect
+#'   # * all _creating_ statements should appear immediately
+#'   # * the first two `future_promise()` calls should start immediately
+#'   # * two `future_promise()` calls should execute simultaneously throughout the process
+#'   # * the log should have `print_every()` statements interleaved with `future_promise()` statements
+#'   {
+#'     log <- tempfile(); start <- Sys.time()
+#'     print_every()
+#'     lapply(1:10, function(i) {
+#'       cat_ex("future_promise(): ", i, " creating")
+#'       future_promise({
+#'         cat_ex("future_promise(): ", i, " starting")
+#'         Sys.sleep(1)
+#'         cat_ex("future_promise(): ", i, " finished")
+#'         i
+#'       })
+#'     }) %>%
+#'       print_ex()
+#'     cat_ex("done with lapply()")
+#'
+#'     run_ex_now() # ~5s; max(5s for `future_promise()`, 5s for `print_every()`)
```

Comment:
Consider wrapping all this template code into a reusable function (to help surface the difference)